File: algorithms/code/leetcode/lc100_same_tree/LC100SameTree.java
package leetcode.lc100_same_tree;

import java.util.ArrayList;

/**
 * https://leetcode.com/problems/same-tree/ #easy
 */
public final class LC100SameTree {
    public static class TreeNode {
        public int val;
        public TreeNode left;
        public TreeNode right;

        TreeNode(final int x) {
            val = x;
        }
    }

    private static class TreeNodePair {
        public TreeNode node1;
        public TreeNode node2;

        TreeNodePair(final TreeNode node1Value, final TreeNode node2Value) {
            node1 = node1Value;
            node2 = node2Value;
        }
    }

    private enum NullCheckResult {
        ALL_NULL, SOME_NULL, NONE_NULL
    }

    private static NullCheckResult nullCheck(final TreeNode tree1, final TreeNode tree2) {
        boolean allNull = tree1 == null && tree2 == null;
        boolean noneNull = tree1 != null && tree2 != null;
        boolean someNull = !allNull && !noneNull;
        if (allNull) {
            return NullCheckResult.ALL_NULL;
        }
        if (someNull) {
            return NullCheckResult.SOME_NULL;
        }
        return NullCheckResult.NONE_NULL;
    }

    public boolean isSameTree(final TreeNode tree1, final TreeNode tree2) {
        NullCheckResult check = nullCheck(tree1, tree2);
        if (check != NullCheckResult.NONE_NULL) {
            return check == NullCheckResult.ALL_NULL;
        }
        ArrayList<TreeNodePair> queue = new ArrayList<>();
        queue.add(new TreeNodePair(tree1, tree2));
        while (queue.size() > 0) {
            TreeNodePair pair = queue.remove(0);
            TreeNode node1 = pair.node1;
            TreeNode node2 = pair.node2;
            if (node1.val != node2.val) {
                return false;
            }
            NullCheckResult checkLeft = nullCheck(node1.left, node2.left);
            if (checkLeft == NullCheckResult.NONE_NULL) {
                queue.add(new TreeNodePair(node1.left, node2.left));
            } else {
                if (checkLeft == NullCheckResult.SOME_NULL) {
                    return false;
                }
            }
            NullCheckResult checkRight = nullCheck(node1.right, node2.right);
            if (checkRight == NullCheckResult.NONE_NULL) {
                queue.add(new TreeNodePair(node1.right, node2.right));
            } else {
                if (checkRight == NullCheckResult.SOME_NULL) {
                    return false;
                }
            }
        }
        return true;
    }
}
package leetcode.lc100_same_tree;

import org.junit.jupiter.api.Test;

import leetcode.lc100_same_tree.LC100SameTree.TreeNode;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

public final class LC100SameTreeTests {
    private TreeNode exampleTree() {
        TreeNode tree0 = new TreeNode(0);
        TreeNode tree1 = new TreeNode(1);
        TreeNode tree2 = new TreeNode(2);
        tree0.left = tree1;
        tree0.right = tree2;
        tree1.left = null;
        tree2.right = null;
        return tree0;
    }

    @Test
    public void testDifferent() throws Exception {
        TreeNode an0 = exampleTree();
        TreeNode bn0 = exampleTree();
        bn0.right.val = -bn0.right.val;
        assertFalse(new LC100SameTree().isSameTree(an0, bn0));
    }

    @Test
    public void testSame() throws Exception {
        TreeNode an0 = exampleTree();
        TreeNode bn0 = exampleTree();
        assertTrue(new LC100SameTree().isSameTree(an0, bn0));
    }

    @Test
    public void testOneEmpty() throws Exception {
        TreeNode tree = new TreeNode(0);
        assertFalse(new LC100SameTree().isSameTree(tree, null));
        assertFalse(new LC100SameTree().isSameTree(null, tree));
    }

    @Test
    public void testBothEmpty() throws Exception {
        assertTrue(new LC100SameTree().isSameTree(null, null));
    }

    @Test
    public void testLeft() throws Exception {
        TreeNode an0 = new TreeNode(0);
        TreeNode bn0 = new TreeNode(0);
        TreeNode an1 = new TreeNode(1);
        an0.left = an1;
        assertFalse(new LC100SameTree().isSameTree(an0, bn0));
    }

    @Test
    public void testRight() throws Exception {
        TreeNode an0 = new TreeNode(0);
        TreeNode bn0 = new TreeNode(0);
        TreeNode an1 = new TreeNode(1);
        an0.right = an1;
        assertFalse(new LC100SameTree().isSameTree(an0, bn0));
    }
}

File: algorithms/code/leetcode/lc100_same_tree/lc100_same_tree.rs
// https://leetcode.com/problems/same-tree/
// #easy

use std::cell::RefCell;
use std::collections::VecDeque;
use std::rc::Rc;

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    pub fn new(
        val: i32,
        left: Option<Rc<RefCell<TreeNode>>>,
        right: Option<Rc<RefCell<TreeNode>>>,
    ) -> Option<Rc<RefCell<Self>>> {
        Some(Rc::new(RefCell::new(TreeNode { val, left, right })))
    }
}

struct TreeNodePair {
    node1: Rc<RefCell<TreeNode>>,
    node2: Rc<RefCell<TreeNode>>,
}

#[derive(PartialEq)]
enum NullCheckResult {
    All,
    Some,
    None,
}

fn null_check(
    tree1: &Option<Rc<RefCell<TreeNode>>>,
    tree2: &Option<Rc<RefCell<TreeNode>>>,
) -> NullCheckResult {
    match (tree1, tree2) {
        (None, None) => NullCheckResult::All,
        (Some(_), Some(_)) => NullCheckResult::None,
        _ => NullCheckResult::Some,
    }
}

pub struct Solution;

impl Solution {
    pub fn is_same_tree(
        tree1: Option<Rc<RefCell<TreeNode>>>,
        tree2: Option<Rc<RefCell<TreeNode>>>,
    ) -> bool {
        let check = null_check(&tree1, &tree2);
        if check != NullCheckResult::None {
            return check == NullCheckResult::All;
        }
        let mut queue = VecDeque::new();
        queue.push_back(TreeNodePair {
            node1: tree1.unwrap(),
            node2: tree2.unwrap(),
        });
        while let Some(pair) = queue.pop_front() {
            let node1 = pair.node1.borrow();
            let node2 = pair.node2.borrow();
            if node1.val != node2.val {
                return false;
            }
            let check_left = null_check(&node1.left, &node2.left);
            if check_left == NullCheckResult::None {
                queue.push_back(TreeNodePair {
                    node1: node1.left.as_ref().unwrap().clone(),
                    node2: node2.left.as_ref().unwrap().clone(),
                });
            } else if check_left == NullCheckResult::Some {
                return false;
            }
            let check_right = null_check(&node1.right, &node2.right);
            if check_right == NullCheckResult::None {
                queue.push_back(TreeNodePair {
                    node1: node1.right.as_ref().unwrap().clone(),
                    node2: node2.right.as_ref().unwrap().clone(),
                });
            } else if check_right == NullCheckResult::Some {
                return false;
            }
        }
        true
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_different() {
        let an0_left = TreeNode::new(1, None, None);
        let an0_right = TreeNode::new(2, None, None);
        let an0 = TreeNode::new(0, an0_left, an0_right);
        let bn0_left = TreeNode::new(1, None, None);
        let bn0_right = TreeNode::new(-2, None, None);
        let bn0 = TreeNode::new(0, bn0_left, bn0_right);
        assert!(!Solution::is_same_tree(an0, bn0));
    }

    #[test]
    fn test_same() {
        let an0_left = TreeNode::new(1, None, None);
        let an0_right = TreeNode::new(2, None, None);
        let an0 = TreeNode::new(0, an0_left, an0_right);
        let bn0_left = TreeNode::new(1, None, None);
        let bn0_right = TreeNode::new(2, None, None);
        let bn0 = TreeNode::new(0, bn0_left, bn0_right);
        assert!(Solution::is_same_tree(an0, bn0));
    }

    #[test]
    fn test_one_empty() {
        let tree = TreeNode::new(0, None, None);
        assert!(!Solution::is_same_tree(tree.clone(), None));
        assert!(!Solution::is_same_tree(None, tree));
    }

    #[test]
    fn test_both_empty() {
        assert!(Solution::is_same_tree(None, None));
    }

    #[test]
    fn test_left() {
        let an1 = TreeNode::new(1, None, None);
        let an0 = TreeNode::new(0, an1, None);
        let bn0 = TreeNode::new(0, None, None);
        assert!(!Solution::is_same_tree(an0, bn0));
    }

    #[test]
    fn test_right() {
        let an1 = TreeNode::new(1, None, None);
        let an0 = TreeNode::new(0, None, an1);
        let bn0 = TreeNode::new(0, None, None);
        assert!(!Solution::is_same_tree(an0, bn0));
    }
}
